compysition/actor.py

```diff
@@ -258,34 +295,42 @@ def __do_consume(self, function, event, queue):
                 event = new_event
 
             if self.REQUIRED_EVENT_ATTRIBUTES:
-                missing = [event.get(attribute) for attribute in self.REQUIRED_EVENT_ATTRIBUTES if not event.get(attribute, None)]
+                missing = [attribute for attribute in self.REQUIRED_EVENT_ATTRIBUTES if not event.get(attribute, None)]
                 if len(missing) > 0:
                     raise InvalidActorInput("Required incoming event attributes were missing: {missing}".format(missing=missing))
 
-            function(event, origin=queue.name, origin_queue=queue)
-        except QueueFull as err:
-            err.wait_until_free()
-            queue.put(event)
+            try:
+                function(event, origin=queue.name, origin_queue=queue)
+            except QueueFull as err:
+                err.queue.wait_until_free() # potential TypeError if target queue is not sent
+                queue.put(event) # puts event back into origin queue
         except InvalidActorInput as error:
             self.logger.error("Invalid input detected: {0}".format(error))
         except InvalidEventConversion:
             self.logger.error("Event was of type '{_type}', expected '{input}'".format(_type=type(event), input=self.input))
         except Exception as err:
             self.logger.warning("Event exception caught: {traceback}".format(traceback=traceback.format_exc()), event=event)
-            rescue_tracker = "{actor}_rescue_num".format(actor=self.name)
-            if self.rescue and event.get(rescue_tracker, 0) < self.max_rescue:
-                setattr(event, rescue_tracker, event.get(rescue_tracker, 0) + 1)
+            rescue_attribute = Actor._RESCUE_ATTRIBUTE_NAME_TEMPLATE.format(actor=self.name)
+            rescue_attempts =  event.get(rescue_attribute, 0)
+            if self.rescue and rescue_attempts < self.max_rescue:
+                setattr(event, rescue_attribute, rescue_attempts + 1)
                 sleep(1)
                 queue.put(event)
             else:
                 event.error = err
                 self.send_error(event)
 
     def create_event(self, *args, **kwargs):
-        if len(self.output) == 1:
-            return self.output[0](**kwargs)
-        raise ValueError("Unable to call create_event function with multiple output types defined")
-
+        try:
+            self.output[1]
+        except IndexError:
+            try:
+                return self.output[0](**kwargs)
+            except IndexError:
+                raise ValueError("Unable to call create_event function without an output type defined")
+        else:
+            raise ValueError("Unable to call create_event function with multiple output types defined")
+            
```

Comment:
I guess what's the reason behind changing this function at all? Looks like the way it was before passes the tests you added, and I'm always hesitant to change stuff that works.

Reply:
The difference is the error message which allows for properly identifying the issue.  Is it strictly necessary, no, but then neither are any of the logs/prints.  That said I'm not attached to this change and would be fine without it.

Reply:
That makes sense, I can see how that would be helpful.